rest_api/sawtooth_rest_api/routes.py

```diff
@@ -178,56 +139,60 @@ def _try_validator_request(self, message_type, content):
         try:
             response = future.result(timeout=timeout)
         except FutureTimeoutError:
-            raise web.HTTPServiceUnavailable(reason=timeout_msg)
+            raise errors.ValidatorUnavailable()
 
         return response.content
 
     @staticmethod
-    def _try_response_parse(proto, response):
+    def _try_response_parse(proto, response, traps=None):
         """
         Parses a protobuf response from the validator
         Raises common validator error statuses as HTTP errors
         """
-        unknown_msg = 'An unknown error occured with your request'
-        notfound_msg = 'There is no resource at that root, address or prefix'
-
         parsed = proto()
         parsed.ParseFromString(response)
+        traps = traps or []
 
+        # Not every protobuf has every status enum, so pass AttributeErrors
+        try:
+            traps.append(errors.Unknown(proto.ERROR))
+        except AttributeError:
+            pass
         try:
-            if parsed.status == proto.ERROR:
-                raise web.HTTPInternalServerError(reason=unknown_msg)
-            if parsed.status == proto.NORESOURCE:
-                raise web.HTTPNotFound(reason=notfound_msg)
+            traps.append(errors.MissingGenesis(proto.NOGENESIS))
         except AttributeError:
-            # Not every protobuf has every status, so pass AttributeErrors
             pass
 
+        for trap in traps:
+            trap.check(parsed.status)
+
         return MessageToDict(parsed, preserving_proto_field_name=True)
 
     @staticmethod
-    def _wrap_response(data=None, head=None, link=None):
+    def _safe_get(obj, key, default=''):
+        """
+        aiohttp very helpfully parses param strings to replace '+' with ' '
+        This is very bad when your block ids contain meaningful +'s
+        """
+        return obj.get(key, default).replace(' ', '+')
```

Comment:
It think my suggestion above was to apply that to where it's needed on the output coming from zmq before heading to the client.